meta-store/src/meta_store/single_file.rs

```diff
@@ -0,0 +1,1027 @@
+        let contents: Vec<u8>;
+
+        {
+            let mut data = self.data.write().unwrap();
+
+            match data.splits.get_mut(&split_id) {
+                Some(m) => {
+                    match m.state {
+                        SplitState::Published => {
+                            // If the split is already published, this API call returns a success.
+                            return Ok(());
+                        }
+                        SplitState::Staged => m.state = SplitState::Published,
+                        _ => {
+                            return Err(MetaStoreErrorKind::SplitIsNotStaged.with_error(
+                                anyhow::anyhow!("Split ID is not staged: {:?}", &split_id),
+                            ));
+                        }
+                    }
+                }
+                None => {
+                    return Err(MetaStoreErrorKind::DoesNotExist
+                        .with_error(anyhow::anyhow!("Split ID does not exist: {:?}", &split_id)));
+                }
+            };
+
+            let meta_data_set = MetaDataSet {
+                index: data.index.clone(),
+                splits: data.splits.clone(),
+            };
+
+            // Serialize meta data.
+            contents = match serde_json::to_vec(&meta_data_set) {
+                Ok(c) => c,
+                Err(e) => {
+                    return Err(MetaStoreErrorKind::InvalidManifest
+                        .with_error(anyhow::anyhow!("Failed to serialize meta data: {:?}", e)));
+                }
+            };
+        }
+
+        // Put data back into storage.
+        match self
+            .storage
+            .put(&self.path, PutPayload::from(contents))
+            .await
+        {
+            Ok(_) => (),
+            Err(e) => {
+                return Err(MetaStoreErrorKind::InternalError
+                    .with_error(anyhow::anyhow!("Failed to put meta data: {:?}", e)));
+            }
+        };
+
+        Ok(())
+    }
+
+    async fn list_splits(
+        &self,
+        state: SplitState,
+        time_range: Option<Range<u64>>,
+    ) -> MetaStoreResult<HashMap<SplitId, SplitManifest>> {
+        let mut splits = HashMap::new();
+
+        let meta_data_set = self.data.read().unwrap();
+
+        for (split_id, split_manifest) in &meta_data_set.splits {
+            if split_manifest.state != state {
+                // If the state of the split that stored in meta store is not match with argument,
+                // this split will be ignored.
+                continue;
+            }
+
+            // filter by time range
+            match time_range {
+                Some(ref filter_time_range) => {
+                    match split_manifest.metadata.time_range {
+                        Some(ref split_time_range) => {
+                            if split_time_range.contains(&filter_time_range.start)
```

Comment:
We want to return [2,3,4]. I believe this predicate is going to miss split 3. (hope this renders well)
 ```
           <----------->
 <--1--><--2--><--3--><--4-->
 ````